pkg/cmd/build/build.go

```diff
@@ -0,0 +1,421 @@
+		if quiet {
+			fmt.Fprintln(output, img.Target.Digest)
+		} else {
+			fmt.Fprintf(output, "Loaded image: %s\n", img.Name)
+		}
+	}
+
+	return nil
+}
+
+func generateBuildctlArgs(ctx context.Context, stdin io.Reader, options *types.BuildCommandOptions) (buildCtlBinary string,
+	buildctlArgs []string, needsLoading bool, metaFile string, tags []string, cleanup func(), err error) {
+
+	buildctlBinary, err := buildkitutil.BuildctlBinary()
+	if err != nil {
+		return "", nil, false, "", nil, nil, err
+	}
+
+	buildContext := options.BuildContext
+	output := options.Output
+	namespace := options.GOptions.Namespace
+	address := options.GOptions.Address
+	platform := options.Platform
+	buildkitHost := options.BuildKitHost
+	if output == "" {
+		client, ctx, cancel, err := clientutil.NewClient(ctx, namespace, address)
+		if err != nil {
+			return "", nil, false, "", nil, nil, err
+		}
+		defer cancel()
+		info, err := client.Server(ctx)
+		if err != nil {
+			return "", nil, false, "", nil, nil, err
+		}
+		sharable, err := isImageSharable(buildkitHost, namespace, info.UUID, options.GOptions.Snapshotter, platform)
+		if err != nil {
+			return "", nil, false, "", nil, nil, err
+		}
+		if sharable {
+			output = "type=image,unpack=true" // ensure the target stage is unlazied (needed for any snapshotters)
+		} else {
+			output = "type=docker"
+			if len(platform) > 1 {
+				// For avoiding `error: failed to solve: docker exporter does not currently support exporting manifest lists`
+				// TODO: consider using type=oci for single-platform build too
+				output = "type=oci"
+			}
+			needsLoading = true
+		}
+	} else {
+		if !strings.Contains(output, "type=") {
+			// should accept --output <DIR> as an alias of --output
+			// type=local,dest=<DIR>
+			output = fmt.Sprintf("type=local,dest=%s", output)
+		}
+		if strings.Contains(output, "type=docker") || strings.Contains(output, "type=oci") {
+			needsLoading = true
+		}
+	}
+	if tags = strutil.DedupeStrSlice(options.Tag); len(tags) > 0 {
+		ref := tags[0]
+		named, err := dockerreference.ParseNormalizedNamed(ref)
+		if err != nil {
+			return "", nil, false, "", nil, nil, err
+		}
+		output += ",name=" + dockerreference.TagNameOnly(named).String()
+
+		// pick the first tag and add it to output
+		for idx, tag := range tags {
+			named, err := dockerreference.ParseNormalizedNamed(tag)
+			if err != nil {
+				return "", nil, false, "", nil, nil, err
+			}
+			tags[idx] = dockerreference.TagNameOnly(named).String()
+		}
+	} else if len(tags) == 0 {
+		output = output + ",dangling-name-prefix=<none>"
+	}
+
+	buildctlArgs = buildkitutil.BuildctlBaseArgs(buildkitHost)
+
+	buildctlArgs = append(buildctlArgs, []string{
+		"build",
+		"--progress=" + options.Progress,
+		"--frontend=dockerfile.v0",
+		"--local=context=" + buildContext,
+		"--output=" + output,
+	}...)
+
+	dir := buildContext
+	file := buildkitutil.DefaultDockerfileName
+	if options.File != "" {
+		if options.File == "-" {
+			var err error
```

Comment:
do we need this line `var err error`?

Reply:
It's a special scope trick

if we don't have `var err error`, the code should be

```go
		if options.File == "-" {
			var err error
			dir, err  := buildkitutil.WriteTempDockerfile(stdin)
			if err != nil {
				return "", nil, false, "", nil, nil, err
			}
			cleanup = func() {
				os.RemoveAll(dir)
			}
		} else {
			dir, file = filepath.Split(options.File)
		}
```

It means that the new dir variable will be unavailable outside the condition block....